### PR TITLE
Attempt to port KCL v2

### PR DESCRIPTION
**Pull request by trucnguyenlam**
This is my attempt to port the library to KCL v2. This PR seems to fix the problem of workers are requested to shutdown (which is mentioned in #19).

### CODE REVIEW

**Comment by anikifoss** [comment]
@trucnguyenlam Thanks for the Kinesis 2.0 port! I had a quick look through the code, and it look good. My only note so far is that we have different coding styles, so we should add scalafmt with a default config to keep the style the same across the codebase.

Once the styling issues are resolved, I'll do a more thorough review during the next few weeks, and then we can merge this in.

**Comment by trucnguyenlam** [comment]
@lexn82 thanks for your comment, could you please refer me the current project scalafmt config? I will try to set it up to align with the current project style.

**Comment by trucnguyenlam** [comment]
By the way, along with code style, I also need to update the README file (as it is for KCLv1).

File: src/main/scala/com/contxt/kinesis/ConsumerConfig.scala [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        case StreamPositionHorizon =>
+          InitialPositionInStreamExtended.newInitialPosition(
+            InitialPositionInStream.TRIM_HORIZON)
+        case StreamPositionTimestamp =>
+          InitialPositionInStreamExtended.newInitialPositionAtTimestamp(
+            DateFormat.getInstance().parse(config.getString("position.time")))
+      }
+      .getOrElse(latestPos)
+
+    val kinesisClient = Option(kac).getOrElse(KinesisClientUtil.createKinesisAsyncClient(KinesisAsyncClient.builder()))
```
Comment (trucnguyenlam) [inline src/main/scala/com/contxt/kinesis/ConsumerConfig.scala:100]: aws recommends to use [KinesisClientUtil](https://docs.aws.amazon.com/streams/latest/dev/kcl-migration.html) to create client

File: src/it/scala/com/contxt/kinesis/TestStreamConfig.scala [hunk 2]
```diff
@@ -1,45 +1,61 @@
...
-      workerId
+      applicationName,
+      workerId,
+      kinesisClient,
+      DynamoDbAsyncClient.builder.region(Region.of(regionName)).build(),
+      CloudWatchAsyncClient.builder.region(Region.of(regionName)).build(),
+      initialPositionInStream,
+      coordinatorConfig = Some(new CoordinatorConfig(applicationName).
+        shardConsumerDispatchPollIntervalMillis(idleTimeBetweenGetRecords.toMillis)),
+      retrievalConfig = Some(
```
Comment (trucnguyenlam) [inline src/it/scala/com/contxt/kinesis/TestStreamConfig.scala:50]: create KCL client with disabled enhanced fanout by default

**Comment by trucnguyenlam** [comment]
@lexn82 I have applied a `scalafmt.conf` and changed README file, could you please have a look?

**Comment by anikifoss** [comment]
@trucnguyenlam Thanks for the additional work on your patch and updating the readme! Let's bump the project version to `3.0.0` instead of just `2.1.0` since this is a pretty major change.

I've applied scalafmt with default config to master. If you run the scalafmt with the same settings, we should be in a good place to remove any style diffs. (I've noticed switching scalafmt styles leads to extra formatting, consider rebasing without intellij scalafmt changes, to make any merge conflicts go away.)

**Comment by trucnguyenlam** [comment]
@lexn82 Thanks for putting the default `scalafmt` in master branch, I have resolved the conflict and reapply `scalafmt`. Hopefully the code style will be consistent from now.
I also change version of the project in README to `3.0.0` as suggested (once it is released).

File: README.md [hunk 3]
```diff
@@ -129,6 +107,27 @@ calling `markProcessed()`, or the steam must be terminated with an exception. If
 after failing to process a record, and not marking it as processed, then no further records can be checkpointed,
 eventually causing the system to run out of memory.
 
+## Consumer Configuration
+The Kinesis Consumer `ConsumerConfig` can be configured via HOCON configuration, which is common for Akka projects
+```hocon
+consumer {
```
Comment (anikifoss) [inline README.md]: `consumer` namespace at the top level may cause collision with other configurations (there may be other libraries doing consumer stuff and configured via HOCON). How about `kinesis-consumer` instead?

File: src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala [hunk 4]
```diff
@@ -45,72 +66,112 @@ object KinesisResourceManager {
...
         case NonFatal(e) => // Ignore.
       }
     }
   }
+
+  private def waitForTableToBecomeAvailable(tableName: String, dynamoDb: DynamoDbClient): Unit = {
+    println("Waiting for " + tableName + " to become ACTIVE...")
+    val startTime = System.currentTimeMillis
+    val endTime = startTime + (10 * 60 * 1000)
+    while ({
```
Comment (anikifoss) [inline src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala]: Could be simplified to one line.

File: src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala [hunk 5]
```diff
@@ -45,72 +66,112 @@ object KinesisResourceManager {
...
+      } catch {
         case NonFatal(e) => // Ignore.
       }
     }
   }
+
+  private def waitForTableToBecomeAvailable(tableName: String, dynamoDb: DynamoDbClient): Unit = {
+    println("Waiting for " + tableName + " to become ACTIVE...")
+    val startTime = System.currentTimeMillis
+    val endTime = startTime + (10 * 60 * 1000)
```
Comment (anikifoss) [inline src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala]: Is DynamoDb really that slow? 10 minute wait time seems excessive.

File: src/main/scala/com/contxt/kinesis/ConsumerConfig.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
+package com.contxt.kinesis
```
Comment (anikifoss) [inline src/main/scala/com/contxt/kinesis/ConsumerConfig.scala:1]: Stopping here for now.

File: src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala (hunk 5 above)
Comment (trucnguyenlam) [reply src/it/scala/com/contxt/kinesis/KinesisResourceManager.scala]: 10minute is the maximum timeout, this loop can terminate early.

**Comment by trucnguyenlam** [comment]
@lexn82: I've applied some small changes, please let me know if you are happy with those.

**Comment by anikifoss** [comment]
@trucnguyenlam Thanks for the update!

I recently had a new addition to the family, so it's been challenging to find the time for coding. Thank you for your patience! Would it be alright with you if I could open a PR against your branch? Small changes are easier to fix in code rather than having the back and forth over the Github messages. That should speed things up a little bit.

**Comment by trucnguyenlam** [comment]
@lexn82 Congratz on your good news! And surely, it's totally ok to open a PR to my branch.

**Comment by bpg-dev** [comment]
Sorry, @lexn82, @trucnguyenlam, I did not notice this PR when was working in a parallel branch to bump scala version to 2.13. That branch have been merged and that caused conflicts here.
I'll fix them and make a PR to trucnguyenlam:port-kcl-v2

**Comment by bpg-dev** [comment]
> I'll fix them and make a PR to trucnguyenlam:port-kcl-v2

Here you go: https://github.com/trucnguyenlam/kcl-akka-stream/pull/1

**Comment by trucnguyenlam** [comment]
@bpg thanks for making a PR to my branch, I'll have a look into that and get it merged ASAP

**Comment by trucnguyenlam** [comment]
@bpg I've merged your PR and resolve the conflict. Please have a look.

**Comment by anikifoss** [comment]
@bpg @agenovese I could not help with this project for the last couple of month. Could you take over and review the PR?

**Comment by agenovese** [comment]
I'll see if I can put together the time, as you can imagine things have been really hectic lately.

My Apologies @trucnguyenlam for the delay, and I want to express my appreciation for the PR. I will do my best to get to this in the next week or two.

**Comment by trucnguyenlam** [comment]
@agenovese Thanks, let's get this done in two weeks 👍

**Review by elise-scx — APPROVED** [review]
Looks good to me! For fun I updated one of my projects to use this version, and it works nicely. Updating does require some extra code changes, so incrementing the major version to 3 was a good call.

**Review by angelo-streetcontxt — APPROVED** [review]
Looks good to me. 

Merging now and will tag and publish a release later today.

I want to say thank you again Truc, and my apologies for how long this took to get merged.

**Comment by trucnguyenlam** [comment]
Thanks a lot for getting this done 👍

**Comment by trucnguyenlam** [comment]
@angelo-streetcontxt by the way, do you know how long this would take (roughly) for the artifact to appear on maven repo [here](https://mvnrepository.com/artifact/com.streetcontxt/kcl-akka-stream)? I haven't uploaded any artifact to maven yet so I'm not clear on that :)

**Comment by angelo-streetcontxt** [comment]
To be honest, I didn't even realize we had it showing up there. As far as I knew, assuming you have our bintray repo setup, the artefact should have become available pretty much immediately.

In our builds we have:

`resolvers in ThisBuild += Resolver.bintrayRepo("streetcontxt", "maven")`

and 

`libraryDependencies += "com.streetcontxt" %% "kcl-akka-stream" % "3.0.0"`

I'll see if I can figure out how/where those additional releases got to a more "official" server

**Comment by angelo-streetcontxt** [comment]
Ok, after a bit of digging, that site must be indexing out repo (the one I mentioned above). 
Not sure how frequently they crawl the repositories, and it doesn't seem to be stated on their site as far as I can tell.

If you're using SBT to download the dependency, the code above should work, if you are downloading it manually you should be able to get the jar here:

https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.11/3.0.0/
https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.12/3.0.0/
https://dl.bintray.com/streetcontxt/maven/com/streetcontxt/kcl-akka-stream_2.13/3.0.0/

**Comment by trucnguyenlam** [comment]
Just tested on your repo
```
resolvers in ThisBuild += Resolver.bintrayRepo("streetcontxt", "maven")
```

works perfectly, thanks 👍

**Pull request merged**